python/llm/src/bigdl/llm/transformers/models/llama.py

```diff
@@ -103,9 +113,15 @@ def llama_attention_forward_4_31(
     kv_seq_len = key_states.shape[-2]
     if past_key_value is not None:
         kv_seq_len += past_key_value[0].shape[-2]
-    cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
-    query_states, key_states = apply_rotary_pos_emb(query_states, key_states,
-                                                    cos, sin, position_ids, "llama")
+    # cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
+    # query_states, key_states = apply_rotary_pos_emb(query_states, key_states,
+    #                                                 cos, sin, position_ids, "llama")
+    q_embed = torch.empty(query_states.shape, dtype=query_states.dtype, device=query_states.device)
+    k_embed = torch.empty(key_states.shape, dtype=key_states.dtype, device=key_states.device)
+    
+    linear_q4_0.apply_rotary_embedding_half_qk(query_states, key_states, position_ids, q_embed, k_embed)
```

Comment:
this only works on GPU, yes?

Reply:
I think cpu ipex has a similar function.

Reply:
> I think cpu ipex has a similar function.

We can add our GPU optimizations first